Libceed scipy paper

valeriabarra opened the pull request:
Automated PR comment:
------------------------------------------
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

My comment:
------------------------------------------
To compile:
```
$ cd scipy_proceedings
```

```
$ ./make_paper.sh papers/libceed-paper/
```

Output in PDF can be found in `scipy_proceedings/output/libceed-paper/paper.pdf`

valeriabarra committed:
Initial commit for libCEED's SciPy2020 paper submission - Add README

valeriabarra committed:
Update README.rst

valeriabarra committed:
Add text from abstract submission

valeriabarra committed:
Have only paper.rst in top-level paper directory as per their instructions

valeriabarra committed:
Use format from conference template

Note: figure labels do not work as shown in their examples https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst

They were removed, otherwise it would not compile

valeriabarra commented:
Update:
I have updated the file according to the conference proceeding template.

I have struggled for quite some time to make figure labels work according to their [example](https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst) but it did not work for me, I don't know why. I removed the labels for now, but of course this way we have undefined references when referring to the figures in the text.

jeremylt commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+   Comparison of memory transfer and floating point operations per degree of freedom for different representations of a linear operator for a PDE in 3D with :math:`b` components and variable coefficients arising due to Newton linearization of a material nonlinearity. The representation labeled as *tensor* computes metric terms on the fly and stores a compact representation of the linearization at ...
+
+In this work, we first introduce libCEED’s conceptual framework and C interface, and then illustrate its new Python interface, developed using the C Foreign Function Interface (CFFI) for Python. CFFI allows to reuse most of the C declarations and requires only a minimal adaptation of some of them. The C and Python APIs are mapped in a nearly 1:1 correspondence. For instance, data stored in the Ce ...
```
I'm not sure how much effort it would be, but we could try using numpy or gnumpy so we have clearer CUDA + Python support? If its easy to add, it would be worth mentioning. I can look into it.

jedbrown replied:
I think JAX is probably the most interesting approach for device arrays, and it would bring with it the best in Python automatic differentiation.

jeremylt replied:
If we can run some tests to verify, then we already have libCEED + Python + CUDA and can say so here, as you can use jax.numpy in place of numpy, if I'm understanding correctly.

jedbrown replied:
FFI aspects of `jax.numpy` are surely different (at least for device arrays).

BTW, JAX is sort of a combined successor to numba and autograd (in the sense that developers moved from both of those projects to JAX); it does both JIT and AD, both with device support. There might be very little work needed to run on `jax.numpy` arrays, but someone would need to experiment.

YohannDudouit commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. class:: abstract
+
+   libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. libCEED offers implementations, selectable at runtime, tuned for a variety of computational device types, including CPUs and GPUs. libCEED’s purely algebraic framework can unobtrusively be integrated in new an ...
```
I feel like we should mention that libCEED is a "Matrix-Free Finite Element library".

jeremylt commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+   A = P^T \underbrace{G^T B^T D B G}_{\text{libCEED}} P \, ,
+
+where :math:`P` is the parallel process decomposition operator (external to ``libCEED``) in which the degrees of freedom (DOFs) are scattered to and gathered from the different compute devices. The operator denoted by :math:`A_L = G^T B^T D B G` gives the local action on a compute node or process, where :math:`G` is a local element restriction operation that localizes DOFs based on the elements,  ...
```
Perhaps here it would be appropriate to mention "This parallel decomposition is out of scope for the libCEED API and needs to be managed by the user or an additional package, such as petsc4py."

valeriabarra replied:
We say (external to ``libCEED``) but sure

jeremylt replied:
Right, I mostly was thinking that we could highlight how libCEED can plug in with other packages that are already popular.

YohannDudouit commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+Introduction
+----------------------------------------------------------------------------------
+
+libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. Historically, conventional high-order finite element methods were rarely used for industrial problems because the Jacobian rapidly loses sparsity as the order is increased, leading to unaffordable solve times and ...
```
I find the argument about Jacobians a bit too restrictive, I think that in general high-order impacts the sparsity (if we don't assume any specific basis functions)

YohannDudouit replied:
We could also cite Orszag's paper with Jed's paper.

valeriabarra replied:
I think it is neater this way. One citation specifically refers to the sparsity and one to the cost.

YohannDudouit commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+----------------------------------------------------------------------------------
+
+libCEED is a new open-source, lightweight library designed to leverage the next generation exascale machines by allowing a wide variety of applications to share highly optimized kernels. Historically, conventional high-order finite element methods were rarely used for industrial problems because the Jacobian rapidly loses sparsity as the order is increased, leading to unaffordable solve times and ...
+stability properties offered by higher order discretizations. Nowadays, high-order numerical methods are widely used in Partial Differential Equation (PDE) solvers, but software packages that provide high-performance implementations have often been special-purpose and intrusive. In contrast, libCEED is light-weight, minimally intrusive, and very versatile. When high-order finite/spectral element  ...
```
Maybe we should cite Orszag's paper: "Spectral Methods for Problems in Complex Geometries" as a justification that sparse matrices are not a good representation for high-order.

YohannDudouit replied:
We might also want to mention that algebraic decomposition is more general than a matrix representation of the finite element operator, something like:
This decomposition also presents the benefit that it can represent non-linear finite element operators as well.

jeremylt commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+   Operator decomposition.
+
+The mathematical formulation of QFunctions, described in weak form, is fully separated from the parallelization and meshing concerns. In fact, QFunctions, which can either be defined by the user or selected from a gallery of available built-in functions in the library, are pointwise functions that do not depend on element resolution, topology, or basis degree (selectable at run time). This easily ...
```
Perhaps
"Currently, user-defined QFunctions are written in C and must be precompiled as a foreign function library and loaded via ctypes."
and (if we show easy GPU support, which I think would be cool)
"However, these C QFunctions allow users to compute on GPU devices that libCEED supports."
(If OCCA is merged by then, that will be Nvidia + AMD)

Aside - Numba won't offer single source though, right? You can run Numba + CUDA, but it won't work with our CUDA backends.

valeriabarra replied:
Attempted in [471a5cc](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/471a5cc68c77240ad12631faa9ccb5c17c6e93cd) (altough the diff is hard to read, even though I used a single line per paragraph, as discussed in the documentation PR)

valeriabarra replied:
I'm not sure about your aside to be honest. Need to learn more

jeremylt replied:
Right now, the CUDA/MAGMA backends take the single source QFunctions in C source files and compile them as device kernels, with appropriate change to the CEED_QFUNCTION macro. If I understand the capabilities correctly, we can use Numba to make a function pointer for the CPU backends but that code won't be compatible with single source QFunctions. We might be able to make a backend that works with numba CUDA, though I think it would take more work than on the CPU side as having the QFunction source is more integral to the CUDA backends.

valeriabarra replied:
I see...

jedbrown replied:
We should refer to JAX instead of (or in addition to) Numba. Notes here say sharing of device arrays is supported. https://jax.readthedocs.io/en/latest/CHANGELOG.html?highlight=__cuda_array_interface__#jax-0-1-59-february-11-2020

I believe the cuda-shared and magma backends would be able to use JAX or Numba qfunctions, but cuda-gen would not because the *.cu source is not available (so it would at least need significant JIT shenanigans, if supportable at all).

valeriabarra replied:
Mentioned both possibilities in [c5a9041](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/c5a9041c8f26486f687aaad0b68c75bde90ed197)

jeremylt replied:
/gpu/cuda/ref /gpu/cuda/reg and /gpu/cuda/shared would still require some modification to replace the compilation of the QF with getting the kernel from JAX/Numba but it looks easier. I agree that I'm not sure if /cuda/gen is possible.

YohannDudouit commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+libCEED's API
+----------------------------------------------------------------------------------
+
+LibCEED's API provides the local action of the linear or nonlinear operator without assembling its sparse representation. Let us define the global operator as
```
I find the word "local" confusing here, I would reword like this:
"LibCEED's API provides the action of the linear or nonlinear Finite Element operator without assembling its sparse representation."

YohannDudouit replied:
I think it would be a bit better to give the intuition that instead of forming a single operator using a sparse matrix, we're going to use a product of operations to decompose our operator that will be applied __sequentially__.

valeriabarra replied:
Attempted in [8187f04](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/8187f0446ea4ca87e782424c4e7d6d0514e9a246) please let me know if it sounds good

jeremylt replied:
We perhaps don't want to say the word sequentially? The exact implementation is backend specific. In some cases we apply suboperators partially, like in the CPU backends. What about:

Instead of forming a single operator using a sparse matrix representation, libCEED composes the subcomponents described in equation (:ref:`eq-operator-decomposition`) to provide the action of the operator in a fashion tuned for the hardware available to the user.

Or something like that?

valeriabarra replied:
Attempted again in [204b47d](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/204b47d14efbdf3e04cfdcd0f957a72c2c855eec)

jeremylt replied:
I like it

YohannDudouit replied:
All the implementations are sequential, but are more or less local.

jeremylt replied:
Sort of?
In the blocked CPU backends, some restrictions are fully applied and some are partially applied.
Then in a loop over element batches, active restrictions are applied to a portion of the inputs, the basis operations are applied to a portion of the elements, qfunction is applied for a portion of the elements, the output basis operations are applied to a portion of the elements, and output restrictions are applied to a portion of the elements.

I personally wouldn't call that a sequential application of objects in the decomposition in the way the that all restrictions, all basis operations, a qfunction call, all transpose operations, and all transpose restrictions are applied fully in sequence in non-optimized backends.

YohannDudouit replied:
I wrote "more or less local" the operations are still applied sequentially...

YohannDudouit commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. class:: keywords
+
+   High-performance Python, performance portability, scalability, parallelism
```
Don't we want "Finite Element" and "High order" in the keywords?

valeriabarra replied:
Thanks. I retrieved the ones used on easychair for the abstract submission. I don't recall now if there was a list to choose from. I think we could add some more, and hopefully the reviewers won't complain if they do not exactly match the abstract submission ones.

valeriabarra committed:
Mention petsc4py as an example of external package for parallel decomposition

valeriabarra committed:
Add Orszag citation and some edits to abstract and Intro

valeriabarra committed:
Better describe decomposition and action of operator

valeriabarra committed:
Better explain GPU support in our currently available C QFunctions

valeriabarra committed:
Mention both Numba and Jax to possibly use for Python QFunctions

valeriabarra committed:
Another attempt at describing the decomposition and action of operator

jeremylt commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,72 @@
+
+   Operator decomposition.
+
+The mathematical formulation of QFunctions, described in weak form, is fully separated from the parallelization and meshing concerns. In fact, QFunctions, which can either be defined by the user or selected from a gallery of available built-in functions in the library, are pointwise functions that do not depend on element resolution, topology, or basis degree (selectable at run time). This easily ...
```
I think JAX is all caps

valeriabarra replied:
True. Thanks

valeriabarra committed:
Python -> Python3 in make_paper.sh

This is needed on Ubuntu, hopefully it won't screw anything on your systems, if that is your default Python

valeriabarra committed:
Introduce Numba CUDA array interface for device memory handling

valeriabarra committed:
Expand intro and add sketch of libCEED's role with different backends and user codes

valeriabarra commented:
I've been trying to have `:label:` work in figure captions as they have in their [example](https://github.com/scipy-conference/scipy_proceedings/blob/2020/papers/00_bibderwalt/00_bibderwalt.rst) but I don't know if it's a problem of the `python3-docutils` package for my distribution or something else. But whenever I try to add a figure `:label:` and compile (for instance `` :label:`fig-assembledVsmatrixfree` ``, I get the following error:

```
! Argument of \@gobble has an extra }.
<inserted text> 
                \par 
l.170 ...DUrole{label}{fig-assembledVsmatrixfree}}

```
Please let me know if any of you experience a similar issue.

valeriabarra commented:
If, instead I use the simple referencing in `rst`, that is, the `.. _fig-label:` before the `.. figure::` directive, the compilation completes well and creates a PDF, but the figure has the missing reference anyways.

valeriabarra committed:
Add figure labels and reference them as Fig. in the text

valeriabarra commented:
Hooray! I was finally able to solve my `docutils` distribution issue and now I am able to add and see all figure references. The only thing left that I see not rendered properly now is the `---` for em dash

valeriabarra committed:
Add descriptions of T-vec, L-vec, E-vec and Q-vec and code block examepls

valeriabarra committed:
Add references for website citations

valeriabarra commented:
Hey @jedbrown @YohannDudouit @jeremylt  I have been adding some better descriptions and code snippets of the Python interface. 

Can anyone give me any input on how to proceed? The deadline is in 2 days. We need to merge this here before being able to create the PR on the conference website for the submission on Friday. Many thanks to all.

jeremylt commented on papers/libceed-paper/paper.rst:
```diff
@@ -0,0 +1,258 @@
+.. figure:: libCEEDBackends.png
+
+   The role of libCEED as a lightweight, portable library which provides a low-level API for efficient, specialized implementations. libCEED allows different applications to share highly optimized discretization kernels. :label:`fig-libCEED-backends`
+
```
I think we can drop " and C interface" part of the first sentence since we don't really discuss the C API past mentioning that the Python interface is a wrapper for the C library. That would help address the reviewers that were concerned that this talks too much about C.

valeriabarra replied:
Reworded. I meant we'd introduce the API.

valeriabarra committed:
Reword: interface -> API

valeriabarra committed:
Add Conclusions and Acknowledgments

valeriabarra committed:
Add CEED MS citations

valeriabarra committed:
Add libCEED dev site citation

jedbrown committed:
Trim figure so it doesn't overlap adjacent column

jedbrown committed:
Revise abstract

jedbrown committed:
Spell-check

jedbrown committed:
Copy editing/todos

valeriabarra commented on papers/libceed-paper/paper.rst:
```diff
@@ -34,16 +34,16 @@ High-performance operator evaluations with ease of use: libCEED's Python interfa
 ----------------------------------------------------------------------------------
 
-Historically, conventional high-order finite element methods were rarely used for industrial problems because the matrix describing the action of the operator loses sparsity as the order is increased :cite:`Orszag1980`, leading to unaffordable solve times and memory requirements :cite:`brown2010`. This effect typically limited the order of accuracy to at most quadratic, especially because they ar ...
+Conventional high-order finite element methods have seen very limited use for industrial problems because the matrix describing the action of the operator loses sparsity as the order is increased :cite:`Orszag1980`, leading to unaffordable solve times and memory requirements :cite:`brown2010`. This effect typically limited the polynomial basis degree to at most quadratic, especially because they  ...
```
Thanks @jedbrown I don't understand one of your questions. What did you mean when you said `[TODO: did you mean to say it gave up a lot?] ` Did you refer to the fact that quadratic elements are advantageous in terms of FLOPs per DOFs as explained in the sentence?

jedbrown replied:
It limits the degree **because** "they are computationally advantageous"? Something is missing/unclear in this sentence.

valeriabarra replied:
Thanks. I think I addressed it in [d57be80](https://github.com/valeriabarra/scipy_proceedings/pull/1/commits/d57be804c51c6ff69e303455465d90f31c8fc630)

valeriabarra committed:
Add Lottes-Fischer reference

valeriabarra committed:
Final edits

valeriabarra commented:
Thanks you all for your inputs and suggestions. I made some final edits. 

Please let me know if we can merge this and then I can create the PR in the conference's repo.

jedbrown committed:
Editing...

jedbrown committed:
Editing

valeriabarra committed:
Remove two CEED Technical Report references

valeriabarra committed:
Some edits suggested by @jedbrown

Moved CEED software distribution citations to last sentence before Conclusions

Mentioned JIT together with device Numba/JAX implementations

Rephrased beginning of libCEED's API section

Also changed one instance of flops -> FLOPs

valeriabarra committed:
Add :corresponding: field
